presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java

```diff
@@ -13,32 +13,50 @@
  */
 package com.facebook.presto.sql.tree;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
+import static com.facebook.presto.sql.QueryUtil.simpleQuery;
 import static com.google.common.base.MoreObjects.toStringHelper;
-import static java.util.Collections.emptyList;
 
 public class ShowStats
         extends Statement
 {
-    private final QualifiedName table;
+    private final Query query;
 
+    @VisibleForTesting
     public ShowStats(QualifiedName table)
     {
         this(Optional.empty(), table);
     }
 
-    public ShowStats(NodeLocation location, QualifiedName table)
+    @VisibleForTesting
+    public ShowStats(Query query)
+    {
+        this(Optional.empty(), query);
+    }
+
+    public ShowStats(Optional<NodeLocation> location, QualifiedName table)
     {
-        this(Optional.of(location), table);
+        this(location, createFakeQuery(location, table));
     }
 
-    private ShowStats(Optional<NodeLocation> location, QualifiedName table)
+    public ShowStats(Optional<NodeLocation> location, Query query)
     {
         super(location);
-        this.table = table;
+        this.query = query;
+    }
+
+    private static Query createFakeQuery(Optional<NodeLocation> location, QualifiedName name)
```

Comment:
As far as I understand, using Relation in ShowStats would simplify this part and allow to get rid of this hacky function. I agree that it's added value, but on the other hand this would additionally complicate ShowStatsRewrite as we would have to cover both cases separately there. Currently this code can be simpler as it's only handling generic Query case.
I agree that this would probably be less hacky solution but on the other hand current code is removing at least some complexity from already complex ShowStatsRewriter. Also I would interpret that as `SHOW STATS FOR table` being just a syntactic sugar for `SHOW STATS FOR (SELECT * FROM table)`.

TLDR: This change can be done, but my personal opinion is that it's only moving complexity from one point to the other and for that reason it's probably not worth additional work.